File: app/src/main/java/com/example/lovereminder/DiaryDao.java
package com.example.lovereminder;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

@Dao
public interface DiaryDao {
    @Query("select * from diaries order by date desc")
//    LiveData<List<Diary>> getAllDiaries();
    Observable<List<Diary>> getAllDiaries();

    @Insert
    Completable insertDiary(Diary diary);

    @Delete
    Completable deleteDiary(Diary diary);

    @Update
    Completable updateDiary(Diary diary);

    @Query("select * from diaries where id = :id")
//    LiveData<Diary> getDiaryById(int id);
//    Observable<Diary> getDiaryById (int id);
    Single<Diary> getDiaryById(int id);
}

File: app/src/main/java/com/phucnguyen/lovereminder/database/DiaryDao.kt
package com.phucnguyen.lovereminder.database

import androidx.room.*
import com.phucnguyen.lovereminder.model.Diary
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import kotlinx.coroutines.flow.Flow

@Dao
interface DiaryDao {
    @Query("select * from diaries order by date desc")
    fun findAllDiaries(): Flow<List<Diary>>

    @Insert
    suspend fun insertDiary(diary: Diary): Long

    @Delete
    suspend fun deleteDiary(diary: Diary): Int

    @Update
    suspend fun updateDiary(diary: Diary): Int

    @Query("select * from diaries where id = :id")
    fun findById(id: Int): Flow<Diary>

    @Query("select * from diaries join diaries_fts on diaries_fts.content = diaries.content where diaries_fts match :query")
    fun getDiariesBasedOnString(query: String): Single<List<Diary>>
}